cmd/compose/compose.go

```diff
@@ -466,13 +474,26 @@ func RootCommand(dockerCli command.Cli, backend api.Service) *cobra.Command { //
 				parallel = i
 			}
 			if parallel > 0 {
+				logrus.Debugf("Limiting max concurrency to %d jobs", parallel)
 				backend.MaxConcurrency(parallel)
 			}
-			ctx, err := backend.DryRunMode(cmd.Context(), dryRun)
+
+			// (5) dry run detection
+			ctx, err = backend.DryRunMode(ctx, dryRun)
 			if err != nil {
 				return err
 			}
 			cmd.SetContext(ctx)
+
+			// (6) Desktop integration
+			if db, ok := backend.(desktop.IntegrationService); ok {
```

Comment:
I wish we refactor this init sequence so we don't mutate `backend` but use setup builder according to options:
```
backend := compose.NewComposeService(dockerCli, WithDryRun(xx), WithDesktopIntegration(xx))
```

Reply:
Agreed 😞 the init is overly complicated, especially because of the split between Docker/CLI/plugins, Compose itself, and Cobra 🐍 

This was already kind of hacky - I didn't want to add this to `api.Service` but the concrete `composeService` implementation is unexported

I'll see if I can refactor this in another PR to swap to options an consolidate some things, e.g. we have several lazy/memoized components but they all happen in different ways currently

Reply:
> I didn't want to add this to api.Service

why not ?